dashboard/utils.py

```diff
@@ -0,0 +1,180 @@
+            self.program.title
+        )
+
+    def _get_program_fa_prices(self):
+        """
+        Returns the financial aid possible cost range.
+        """
+        course_max_price = self.program.get_course_price()
+        # get all the possible discounts for the program
+        program_tiers_qset = TierProgram.objects.filter(
+            Q(program=self.program) & Q(current=True)).order_by('discount_amount')
+        if not program_tiers_qset.exists():
+            return course_max_price, course_max_price
+        min_discount = program_tiers_qset.aggregate(
+            Min('discount_amount')).get('discount_amount__min', 0)
+        max_discount = program_tiers_qset.aggregate(
+            Max('discount_amount')).get('discount_amount__max', 0)
+        return course_max_price - max_discount, course_max_price - min_discount
+
+    def _is_course_in_program(self, course_id):
+        """
+        Returns whether the course id belongs to the program
+        """
+        return course_id in self.course_ids
+
+    def is_enrolled(self, course_id):
+        """
+        Returns whether the user is enrolled at least audit in a course run.
+        """
+        return self._is_course_in_program(course_id) and self.enrollments.is_enrolled_in(course_id)
+
+    def is_enrolled_mmtrack(self, course_id):
+        """
+        Returns whether an used is enrolled mmtrack in a course run.
+        This means if the user is enrolled verified for normal programs
+        or enrolled and paid on micromasters for financial aid ones.
+        """
+        if not self.is_enrolled:
+            return False
+        # normal programs need to have a verified enrollment
+        if not self.financial_aid_available:
+            enrollment = self.enrollments.get_enrollment_for_course(course_id)
+            return enrollment.is_verified
+        # financial aid programs need to have an audit enrollment and a paid entry for the course
+        else:
+            return course_id in self.paid_course_ids
+
+    def has_passed_course(self, course_id):
+        """
+        Returns whether the user has passed a course run.
+        This means if the user has a verified certificate for normal programs
+        or a current_grade with passed property and the course has ended.
+        """
+        if not self.is_enrolled_mmtrack(course_id):
+            return False
+
+        # for normal programs need to check the certificate
+        if not self.financial_aid_available:
+            return self.certificates.has_verified_cert(course_id)
+        # financial aid programs need to have an audit enrollment,
+        # a current grade with a passed attribute and the course should be ended
+        else:
+            cur_grade = self.current_grades.get_current_grade(course_id)
+            if cur_grade is None:
+                return False
+            course_run_qset = CourseRun.objects.filter(edx_course_key=course_id)
+            if not course_run_qset.exists():
```

Comment:
Can we use `first()` and check it for `None` instead? There's the (tiny) possibility of a race condition since there are two SQL statements here without a transaction